File: python/7_abstraction/beverage.py
from abc import ABC, abstractmethod

class Beverage(ABC):
    def __init__(self, name: str):
        self.name = name

    @abstractmethod
    def prepare(self):
        pass

    def serve(self):
        print(f"Serving {self.name} in a cozy cup ☕")

class Coffee(Beverage):
    def prepare(self):
        print("Grinding beans...")
        print("Pulling espresso...")
        print("Steaming milk & art ❤️")

class Tea(Beverage):
    def prepare(self):
        print("Heating water...")
        print("Steeping...")
        print("Adding honey")

# Demo
latte = Coffee("Latte")
latte.prepare()
latte.serve()

tea = Tea("Herbal Tea")
tea.prepare()
tea.serve()

File: typescript/7_abstraction/beverage.ts
abstract class Beverage {
    constructor(protected name: string) {}

    abstract prepare(): void;

    serve(): void {
        console.log(`Serving ${this.name} in a cozy cup ☕`);
    }
}

class Coffee extends Beverage {
    constructor() {
        super("Latte");
    }

    prepare(): void {
        console.log("Grinding beans...");
        console.log("Pulling espresso...");
        console.log("Steaming milk & art ❤️");
    }
}

class Tea extends Beverage {
    constructor() {
        super("Herbal Tea");
    }

    prepare(): void {
        console.log("Heating water...");
        console.log("Steeping...");
        console.log("Adding honey");
    }
}

// Demo
const latte = new Coffee();
latte.prepare();
latte.serve();